### PR TITLE
feat(optimizer): Convert Overwindow with unbounded frame to Agg + Join

### LINKED ISSUE #9034: Convert OverAgg with unbounded frame to Agg + Join
A common use case of OverAgg is expanding the aggregation result to the source table, we can simply convert the such query to Agg + Join.

```sql
SELECT depname, empno, salary, avg(salary) OVER (PARTITION BY depname) FROM empsalary;
```

```
  depname  | empno | salary |          avg
-----------+-------+--------+-----------------------
 develop   |    11 |   5200 | 5020.0000000000000000
 develop   |     7 |   4200 | 5020.0000000000000000
 develop   |     9 |   4500 | 5020.0000000000000000
 develop   |     8 |   6000 | 5020.0000000000000000
 develop   |    10 |   5200 | 5020.0000000000000000
 personnel |     5 |   3500 | 3700.0000000000000000
 personnel |     2 |   3900 | 3700.0000000000000000
 sales     |     3 |   4800 | 4866.6666666666666667
 sales     |     1 |   5000 | 4866.6666666666666667
 sales     |     4 |   4800 | 4866.6666666666666667
(10 rows)
```

We can convert it to the following query as an optimizer rule:

```sql
WITH depsalary AS (SELECT depname, avg(salary) AS avgsalary FROM empsalary GROUP BY depname) SELECT empsalary.depname, empno, avgsalary FROM empsalary INNER JOIN depsalary USING (depname);
```

There are two limitations:

1. We should check only aggregate functions are called.
2. No frame clause.

### PR DESCRIPTION
**Pull request by Honeta**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

close #9034

Needs to be merged after #9890. Also, aggregate types such as avg() that would be replaced by LogicalAgg::create() are currently not supported and will cause problems, so this PR can only be safely merged after issue #9929 is resolved.

## Checklist For Contributors

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

### CODE REVIEW

File: src/frontend/src/optimizer/rule/over_window_to_agg_and_join_rule.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        let input_len = over_window.input().schema().len();
+        let mut out_fields = (0..input_len).collect_vec();
+        for i in 0..window_functions.len() {
+            out_fields.push(input_len + group_exprs.len() + i);
+        }
+        let common_input = LogicalShare::create(over_window.input());
+        let agg = LogicalAgg::create(select_exprs, group_exprs, None, common_input.clone());
+        if agg.is_err() {
+            return None;
+        }
```
Comment (st1page) [inline src/frontend/src/optimizer/rule/over_window_to_agg_and_join_rule.rs:79]: ditto.

**Review by st1page — COMMENTED** [review]
please add some e2e tests.

File: src/frontend/src/optimizer/rule/over_window_to_agg_and_join_rule.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                let agg_call = AggCall::new(
+                    kind,
+                    window.args.iter().map(|x| x.clone().into()).collect_vec(),
+                    false,
+                    OrderBy::any(),
+                    Condition::true_cond(),
+                );
+                if agg_call.is_err() {
+                    return None;
+                }
```
[the remark this answers is not in the document]
Comment (Honeta) [reply src/frontend/src/optimizer/rule/over_window_to_agg_and_join_rule.rs:63]: I tried it, and the following prompt appears:
`the ? operator can only be used on Options, not Results, in a method that returns Option`

[the remark this answers is not in the document]
Comment (Honeta) [reply src/frontend/src/optimizer/rule/over_window_to_agg_and_join_rule.rs]: Oh, I found out that ok() should be used here.

File: src/frontend/src/optimizer/rule/over_window_to_agg_and_join_rule.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                    false,
+                    OrderBy::any(),
+                    Condition::true_cond(),
+                )
+                .ok()?;
+                select_exprs.push(agg_call.into());
+            } else {
+                return None;
+            }
+        }
```
Comment (richardchien) [inline src/frontend/src/optimizer/rule/over_window_to_agg_and_join_rule.rs:66]:
```suggestion
        for func in window_functions {
            if let WindowFuncKind::Aggregate(kind) = func.kind {
                let agg_call = AggCall::new(
                    kind,
                    func.args.iter().map(|x| x.clone().into()).collect_vec(),
                    false,
                    OrderBy::any(),
                    Condition::true_cond(),
                )
                .ok()?;
                select_exprs.push(agg_call.into());
            } else {
                return None;
            }
        }
```

File: src/frontend/src/optimizer/rule/over_window_to_agg_and_join_rule.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                        Type::Equal,
+                        vec![
+                            x.clone().into(),
+                            InputRef::new(idx + input_len, x.data_type.clone()).into(),
+                        ],
+                    )
+                    .unwrap()
+                    .into(),
+                ))
+            });
```
Comment (richardchien) [inline src/frontend/src/optimizer/rule/over_window_to_agg_and_join_rule.rs:92]:
```suggestion
        let on_clause = window_functions[0].partition_by.iter().enumerate().fold(
            Condition::true_cond(),
            |on_clause, (idx, x)| {
                on_clause.and(Condition::with_expr(
                    FunctionCall::new(
                        Type::Equal,
                        vec![
                            x.clone().into(),
                            InputRef::new(idx + input_len, x.data_type.clone()).into(),
                        ],
                    )
                    .unwrap()
                    .into(),
                ))
            },
        );
```

File: e2e_test/streaming/over_window.slt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+2 3 7 6
+2 4 7 3
+2 5 7 3
+3 1 4 3
+3 2 4 3
+3 3 4 6
+3 3 4 6
+3 4 4 3
+3 5 4 3
+
```
Comment (richardchien) [inline e2e_test/streaming/over_window.slt:40]: Can we test MV updates by adding another round of `insert` and `select`?

File: src/frontend/src/optimizer/rule/over_window_to_agg_and_join_rule.rs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            }
+        }
+
+        let input_len = over_window.input().schema().len();
+        let mut out_fields = (0..input_len).collect_vec();
+        for i in 0..window_functions.len() {
+            out_fields.push(input_len + group_exprs.len() + i);
+        }
+        let common_input = LogicalShare::create(over_window.input());
+        let agg = LogicalAgg::create(select_exprs, group_exprs, None, common_input.clone()).ok()?;
```
Comment (richardchien) [inline src/frontend/src/optimizer/rule/over_window_to_agg_and_join_rule.rs:74]:
```suggestion
        let (agg, ..) =
	            LogicalAgg::create(select_exprs, group_exprs, None, common_input.clone()).ok()?;
```

File: src/frontend/src/optimizer/logical_optimization.rs [hunk 7]
```diff
@@ -536,6 +537,8 @@ impl LogicalOptimizer {
 
         plan = plan.optimize_by_rules(&LIMIT_PUSH_DOWN);
 
+        plan = plan.optimize_by_rules(&DAG_TO_TREE);
```
[the remark this answers is not in the document]
Comment (Honeta) [reply src/frontend/src/optimizer/logical_optimization.rs:540]: I introduced the DAG in this rule, but if the DAG is not split at the end of gen_optimized_logical_plan_for_batch(), an error will be reported: batch query doesn't support share operator for now.

**Pull request merged**